File: Objective-C/NendIronSourceCustomAdapterObjectiveCSample/NendIronSourceCustomAdapterObjectiveCSample/ViewController.m
//
//  ViewController.m
//  NendIronSourceCustomAdapterObjectiveCSample
//
//  Copyright(c) FAN Communications, Inc. All Rights Reserved
//

#import "ViewController.h"
#import <IronSource/IronSource.h>
@import AppTrackingTransparency;

#define APPKEY @"YOUR_APP_KEY"

@interface ViewController ()

@property (nonatomic) ViewControllerCallback *delegate;
@property (weak, nonatomic) IBOutlet UIButton *showRVButton;
@property (weak, nonatomic) IBOutlet UIButton *showISButton;
@property (weak, nonatomic) IBOutlet UIButton *loadISButton;
@property (weak, nonatomic) IBOutlet UILabel  *versionLabel;

@end

@interface ViewControllerCallback () <LevelPlayRewardedVideoDelegate, LevelPlayInterstitialDelegate>

@property (nonatomic) ISPlacementInfo *rvPlacementInfo;
@property (nonatomic) ViewController *viewController;

@end

@implementation ViewControllerCallback

- (id)initWithViewController:(ViewController *)viewController
{
    self = [super init];
    if (self) {
        self.viewController = viewController;
    }
    return self;
}

#pragma mark - Rewarded Video Delegate Functions

/**
 Called after a rewarded video has been clicked.
 @param placementInfo An object that contains the placement's reward name and amount.
 @param adInfo The info of the ad.
 */
- (void)didClick:(ISPlacementInfo *)placementInfo withAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

/**
 Called after a rewarded video has been viewed completely and the user is eligible for a reward.
 @param placementInfo An object that contains the placement's reward name and amount.
 @param adInfo The info of the ad.
 */
- (void)didReceiveRewardForPlacement:(ISPlacementInfo *)placementInfo withAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    self.rvPlacementInfo = placementInfo;
}

/**
 Called after a rewarded video has changed its availability to true.
 @param adInfo The info of the ad.
 */
- (void)hasAvailableAdWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.viewController.showRVButton setEnabled:true];
    });
}

/**
 Called after a rewarded video has changed its availability to false.
 */
- (void)hasNoAvailableAd {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.viewController.showRVButton setEnabled:false];
    });
}

#pragma mark - Common Rewarded and Interstitial Delegate Functions

/**
 Called after a rewarded video has attempted to show but failed.
 Called after an interstitial has attempted to show but failed.
 @param error The reason for the error
 @param adInfo The info of the ad.
 */
- (void)didFailToShowWithError:(NSError *)error andAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s, error:%@",__PRETTY_FUNCTION__,error);
    if ([adInfo.ad_unit isEqualToString:@"rewarded_video"]) {
        NSLog(@"rewarded_video");
    } else if ([adInfo.ad_unit isEqualToString:@"interstitial"]) {
       NSLog(@"interstitial");
   } else {
       NSLog(@"other format");
   }
}

/**
 Called after a rewarded video has been opened.
 Called after an interstitial has been opened.
 @param adInfo The info of the ad.
 */
- (void)didOpenWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s",__PRETTY_FUNCTION__);
    if ([adInfo.ad_unit isEqualToString:@"rewarded_video"]) {
        NSLog(@"rewarded_video");
    } else if ([adInfo.ad_unit isEqualToString:@"interstitial"]) {
       NSLog(@"interstitial");
   } else {
       NSLog(@"other format");
   }
}

/**
 Called after a rewarded video has been dismissed.
 @param adInfo The info of the ad.
 */
- (void)didCloseWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    NSString *message = [NSString stringWithFormat:@"You have been rewarded %@ %@", self.rvPlacementInfo.rewardAmount, self.rvPlacementInfo.rewardName];
    if ([adInfo.ad_unit isEqualToString:@"rewarded_video"]) {
        if (self.rvPlacementInfo) {
            UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Video Reward"
                                                                                     message:message
                                                                              preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK"
                                                             style:UIAlertActionStyleCancel
                                                           handler:nil];
            [alertController addAction:action];
            [self.viewController presentViewController:alertController
                               animated:YES
                             completion:nil];
            self.rvPlacementInfo = nil;
        }
        NSLog(@"rewarded_video");
    } else if ([adInfo.ad_unit isEqualToString:@"interstitial"]) {
        NSLog(@"interstitial");
    } else {
        NSLog(@"other format");
    }
}

#pragma mark - Interstitial Delegate Functions

/**
 Called after an interstitial has been clicked.
 @param adInfo The info of the ad.
 */
- (void)didClickWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s",__PRETTY_FUNCTION__);
}

/**
 Called after an interstitial has attempted to load but failed.
 @param error The reason for the error
 */
- (void)didFailToLoadWithError:(NSError *)error {
    NSLog(@"%s, error:%@",__PRETTY_FUNCTION__,error);
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.viewController.showISButton setEnabled:false];
    });
}

/**
 Called after an interstitial has been loaded
 @param adInfo The info of the ad.
 */
- (void)didLoadWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s",__PRETTY_FUNCTION__);
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.viewController.showISButton setEnabled:true];
    });
}

- (void)didShowWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s",__PRETTY_FUNCTION__);
}

@end

@implementation ViewController

#pragma mark ViewLifecycle Functions

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.versionLabel.text = [NSString stringWithFormat:@"sdk version %@", [IronSource sdkVersion]];
    self.delegate = [[ViewControllerCallback alloc] initWithViewController:self];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if (@available(iOS 14.0, *)) {
        [self usingATTConsentDialog];
    } else {
        [self setupIronSourceSdk];
    }
}

#pragma mark Private Functions
- (void)usingATTConsentDialog API_AVAILABLE(ios(14.0)) {
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([ATTrackingManager trackingAuthorizationStatus] == ATTrackingManagerAuthorizationStatusNotDetermined) {
            [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
                switch (status) {
                    case ATTrackingManagerAuthorizationStatusAuthorized:
                    case ATTrackingManagerAuthorizationStatusDenied:
                    case ATTrackingManagerAuthorizationStatusNotDetermined:
                    case ATTrackingManagerAuthorizationStatusRestricted:
                        break;
                    default:
                        break;
                }
                [self setupIronSourceSdk];
            }];
        } else {
            [self setupIronSourceSdk];
        }
    });
}

- (void)setupIronSourceSdk {
    [ISIntegrationHelper validateIntegration];
    
    [IronSource setLevelPlayRewardedVideoDelegate:self.delegate];
    [IronSource setLevelPlayInterstitialDelegate:self.delegate];
    
    [IronSource initWithAppKey:APPKEY];
}

#pragma mark Interface Handling

- (IBAction)showRVButtonTapped:(id)sender {
    
    // After calling 'setRVDelegate' and 'initRVWithAppKey:withUserId'
    // you are ready to present an ad. You can supply a placement
    // by calling 'showRVWithPlacementName', or you can simply
    // call 'showRV'. In this case the SDK will use the default
    // placement one created for you.
    [IronSource showRewardedVideoWithViewController:self];
}

- (IBAction)showISButtonTapped:(id)sender {
    
    // This will present the Interstitial. Unlike Rewarded
    // Videos there are no placements.
    [IronSource showInterstitialWithViewController:self];
}

- (IBAction)loadISButtonTapped:(id)sender {
    // This will load the Interstitial. Unlike Rewarded
    // Videos there are no placements.
    [IronSource loadInterstitial];
}

@end

File: Swift/NendIronSourceCustomAdapterSwiftSample/NendIronSourceCustomAdapterSwiftSample/ViewController.swift
//
//  ViewController.swift
//  NendIronSourceCustomAdapterSwiftSample
//
//  Copyright(c) FAN Communications, Inc. All Rights Reserved
//

import UIKit
import Foundation
import ObjectiveC.runtime
import AppTrackingTransparency

let APPKEY = "YOUR_APP_KEY"

class ViewController: UIViewController, ISImpressionDataDelegate {
    
    //MARK: IBOutlets
    @IBOutlet weak var showRVButton: UIButton!
    @IBOutlet weak var showISButton: UIButton!
    @IBOutlet weak var versionLabel: UILabel!
    
    //MARK: IBOutlets Actions
    @IBAction func showRVButton(_ sender: Any) {
        IronSource.showRewardedVideo(with: self)
    }
    
    @IBAction func showISButtonAction(_ sender: Any) {
        IronSource.showInterstitial(with: self)
    }
    
    @IBAction func loadISButtonAction(_ sender: Any) {
        IronSource.loadInterstitial()
    }
    
    var rvPlacementInfo: ISPlacementInfo!
    
    //MARK: ViewLifecycle Functions
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        self.versionLabel.text = String(format: "%@%@", "sdk version: ", IronSource.sdkVersion())
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        if #available(iOS 14, *) {
            usingATTConsentDialog()
        } else {
            self.setupIronSourceSdk()
        }
    }
    
    //MARK: Private Functions
    @available(iOS 14, *)
    private func usingATTConsentDialog() {
        DispatchQueue.main.async {
            if (ATTrackingManager.trackingAuthorizationStatus == .notDetermined) {
                ATTrackingManager.requestTrackingAuthorization { (status) in
                    switch (status) {
                    case.authorized: fallthrough
                    case.denied: fallthrough
                    case.notDetermined: fallthrough
                    case.restricted: fallthrough
                    default:
                        break
                    }
                    self.setupIronSourceSdk()
                }
            } else {
                self.setupIronSourceSdk()
            }
        }
    }
    
    func setupIronSourceSdk() {
        IronSource.setLevelPlayRewardedVideoDelegate(self)
        IronSource.setLevelPlayInterstitialDelegate(self)
        IronSource.add(self)
        
        IronSource.initWithAppKey(APPKEY)
    }
    
    func logFunctionName(string: String = #function) {
        print("Nend ironSource custom adapter Swift Demo App: " + string)
    }
    
    //MARK: ISImpressionData Functions
    func impressionDataDidSucceed(_ impressionData: ISImpressionData!) {
        logFunctionName(string: #function + String(describing: impressionData))
    }
}
    
extension ViewController : LevelPlayRewardedVideoDelegate, LevelPlayInterstitialDelegate {
    //MARK: LevelPlayRewardedVideoDelegate Functions
    /**
     Called after a rewarded video has changed its availability to true.
     @param adInfo The info of the ad.
     */
    func hasAvailableAd(with adInfo: ISAdInfo!) {
        showRVButton.isEnabled = true
        logFunctionName(string: #function)
    }
    
    /**
     Called after a rewarded video has changed its availability to false.
     */
    func hasNoAvailableAd() {
        logFunctionName()
    }
    
    //MARK: LevelPlayRewardedVideoBaseDelegate Functions
    /**
     Called after a rewarded video has been viewed completely and the user is eligible for a reward.
     @param placementInfo An object that contains the placement's reward name and amount.
     @param adInfo The info of the ad.
     */
    func didReceiveReward(forPlacement placementInfo: ISPlacementInfo!, with adInfo: ISAdInfo!) {
        rvPlacementInfo = placementInfo
        logFunctionName(string: #function + String(describing: placementInfo.self))
    }
    
    /**
     Called after a rewarded video has been clicked.
     @param placementInfo An object that contains the placement's reward name and amount.
     @param adInfo The info of the ad.
     */
    func didClick(_ placementInfo: ISPlacementInfo!, with adInfo: ISAdInfo!) {
        logFunctionName(string: #function + String(describing: placementInfo.self))
    }
    
    //MARK: Common LevelPlayRewardedVideoBaseDelegate and LevelPlayInterstitialDelegate Functions
    /**
     Called after a rewarded video has been dismissed.
     Called after an interstitial has been dismissed.
     @param adInfo The info of the ad.
     */
    func didClose(with adInfo: ISAdInfo!) {
        logFunctionName(string: #function)
        if (adInfo.ad_unit == "rewarded_video") {
            print("rewarded_video")
            let message = "You have been rewarded " + rvPlacementInfo.rewardName! + ": \(String(describing: rvPlacementInfo.rewardAmount!))"
            let action = UIAlertAction.init(title: "OK", style: .cancel)
            let alert = UIAlertController.init(title:"Video Reward",
                                               message: message,
                                               preferredStyle: .alert)
            alert.addAction(action)
            self.present(alert, animated: true)
        } else if (adInfo.ad_unit == "interstitial"){
            print("interstitial")
        } else {
            print("other format")
        }
    }
    
    /**
     Called after a rewarded video has been opened.
     Called after an interstitial has been opened.
     @param adInfo The info of the ad.
     */
    func didOpen(with adInfo: ISAdInfo!) {
        logFunctionName(string: #function)
        if (adInfo.ad_unit == "rewarded_video") {
            print("rewarded_video")
        } else if (adInfo.ad_unit == "interstitial"){
            print("interstitial")
        } else {
            print("other format")
        }
    }
    
    /**
     Called after a rewarded video has attempted to show but failed.
     Called after an interstitial has attempted to show but failed.
     @param error The reason for the error
     @param adInfo The info of the ad.
     */
    func didFailToShowWithError(_ error: Error!, andAdInfo adInfo: ISAdInfo!) {
        logFunctionName(string: #function + String(describing: error.self))
        if (adInfo.ad_unit == "rewarded_video") {
            print("rewarded_video")
        } else if (adInfo.ad_unit == "interstitial"){
            showISButton.isEnabled = false
            print("interstitial")
        } else {
            print("other format")
        }
        
    }
    
    //MARK: LevelPlayInterstitialDelegate Functions
    /**
     Called after an interstitial has been clicked.
     @param adInfo The info of the ad.
     */
    func didClick(with adInfo: ISAdInfo!) {
        logFunctionName(string: #function)
    }
    
    /**
     Called after an interstitial has been loaded
     @param adInfo The info of the ad.
     */
    func didLoad(with adInfo: ISAdInfo!) {
        showISButton.isEnabled = true
        logFunctionName(string: #function)
    }
    
    /**
     Called after an interstitial has attempted to load but failed.
     @param error The reason for the error
     */
    func didFailToLoadWithError(_ error: Error!) {
        showISButton.isEnabled = false
        logFunctionName(string: #function)
    }
    
    /**
     Called after an interstitial has been displayed on the screen.
     @param adInfo The info of the ad.
     */
    func didShow(with adInfo: ISAdInfo!) {
        logFunctionName(string: #function)
    }
}
